Docs/component interaction (#7)
paastech-cloud/docs: Architecture.md

- an Application (also referred to as Deployment) is an atomic unit of code, and is the result of a Project deployment. This unit is internally managed and can only be configured to a certain extent by the Client.

### Component interaction

```mermaid
flowchart LR
    cli["CLI"]

    web <-- HTTP --> api
    cli <-- HTTP --> api
    cli <-- Git over SSH --> ssh

    subgraph Github Pages
        web["Web Frontend"]
    end

    subgraph PaaSTech internal services
        api["API"]
        git["Git controller"]
        ssh["OpenSSH server"]
        pom["Pomegranate"]
        db[(PostgreSQL)]
        fs[("`Host filesystem`")]
        docker(["Docker"])
        cr(["`Local container registry`"])

        api -. gRPC .-> git
        api -. gRPC .-> pom
        api -. SQL .-> db

        git -. manages repositories .-> fs
        ssh -. git receive-pack .-> fs
        ssh -. push images .-> cr

        pom -. manages deployments .-> docker

        docker -. use images .-> cr

        cr === fs
    end
```

> In this schema we can see many different component interacting with each other. We decided for this first iteration to split services as much as we could to allow for a more modular architecture. This allows us to easily replace a component if needed, and to scale the different components independently.

Reviewer: ```suggestion
In this schema, we can see the different components interacting with each other. We decided, for this first iteration, to split services as much as we could to allow for a more modular architecture. This allows us to easily replace a component if needed, and to scale the different components independently.
```

> - The client API is the main component that will be used by the clients to interact with the service. This API will then interact with the git controller to create repositories and with pomegranate to create docker deployments, and with the database to store the data.

Reviewer: ```suggestion
- The client API is the main component that will be used by the clients to interact with the service. This API will then interact with the Git controller to create repositories and with Pomegranate to create Docker deployments, and with the database to store the data.
```

> - The git controller will interact with the host filesystem to create repositories, and with the local container registry to push images.

Reviewer: ```suggestion
- The Git controller will interact with the host file system to create repositories, and with the local container registry to push images.
```

> - Pomegranate, our deployment manager, will interact with the local container registry to pull images and with the host filesystem to create deployments.

Reviewer: ```suggestion
- Pomegranate, our deployment manager, will interact with the local container registry to pull images and with the host file system to create deployments.
```

> Most of the interactions between the different components are done through gRPC, except for the web frontend and the CLI, which use HTTP requests. Using gRPC allows for a more efficient communication between the components, as well as a more secure one, as the communication is done through a private network. While the http requests allows us easier access to the API.

Reviewer: ```suggestion
Most of the interactions between the different components are done through gRPC, except for the web frontend and the CLI, which use HTTP requests. Using gRPC allows for a more efficient communication between the components, as well as a more secure one, as the communication is done through a private network. While the HTTP requests allows us easier access to the API.
```
